[TFLite] Move AVERAGE_POOL_3D and MAX_POOL_3D to builtin ops

Tessil opened the pull request:
Hi,

This PR moves the custom `AvgPool3D` and `MaxPool3D` ops to bulitin `AVERAGE_POOL_3D` and `MAX_POOL_3D`  ops (`AVG` is changed to `AVERAGE` for coherence with `AVERAGE_POOL_2D`).

As discussed previously @jianlijianli we tried to keep the code of the custom ops for backward compatibility, notably in [tensorflow/lite/kernels/BUILD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/BUILD#L785) and [tensorflow/lite/kernels/custom_ops_register.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/custom_ops_register.h#L22). Note though that some of the code in [tensorflow/lite/kernels/pooling3d.cc](https://github.com/tensorflow/tensorflow/pull/53206/files#diff-5591b2bdc4d02753edd1722677e7f3e3645a2531a411c904665bb94f3870ec16) for reading the flatbuffer was removed which could create some backward incompatibilities. We could try to keep it in some way if needed but as the custom ops version of these kernels were not compiled into the official binary I'm more in favour of removing this old code if possible along with the one in `tensorflow/lite/kernels/BUILD` and `tensorflow/lite/kernels/custom_ops_register.h`.

Thibaut

Tessil committed:
Move AVERAGE_POOL_3D and MAX_POOL_3D to builtin ops

Tessil committed:
Keep tflite::ops::custom::Register_AVG_POOL_3D for backward compatibility with custom_ops_register.h

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Add QuantizableResult trait to TFL_AveragePool3DOp and TFL_MaxPool3DOp

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Check the zero-point of 3D Pooling ops with TF_LITE_ENSURE instead of TFLITE_DCHECK and adapt the tests to this enforcement

Tessil commented:
@miaout17 As discussed, please see https://arxiv.org/abs/2007.13224 for a network using MaxPool3D and https://keras.io/examples/vision/3D_image_classification/ example that uses it. It'd be nice to be able to convert the Keras model described in the example to TFLite.

Thanks,
Thibaut

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

gbaned commented:
@Tessil Can you please resolve conflicts? Thanks!

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Fix formatting

Tessil commented:
@gbaned Thanks, I fixed the conflicts

miaout17 commented:
Sorry for the wait. I think it looks good to add the 3D pooling ops as builtin. I'm looking for someone to review the code.

sirakiin commented on tensorflow/lite/kernels/pooling3d.cc:
```diff
@@ -442,6 +226,22 @@ TfLiteRegistration* Register_MAX_POOL_3D() {
+
+namespace custom {
+
+// The custom average pooling 3d op registered in custom_ops_register.h uses
```
Shall we also keep Register_MAX_POOL_3D?

Tessil replied:
Effectively, not sure why I didn't initially. Added the implementation for `custom::Register_MAX_POOL_3D`.

sirakiin reviewed (changes requested):
Overall looks good. Thanks for the PR.

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Address builtin AVERGAE/MAX_POOL_3D comments

Tessil commented on tensorflow/compiler/mlir/lite/ir/tfl_ops.td:
```diff
@@ -5517,6 +5517,71 @@ No value constant op.
+  let summary = "Average Pool 3D operator";
+
+  let description = [{
+    Performs average-pooling operation on 3D input.
```
Done

Tessil commented on tensorflow/compiler/mlir/lite/ir/tfl_ops.td:
```diff
@@ -5517,6 +5517,71 @@ No value constant op.
+  let summary = "Max Pool 3D operator";
+
+  let description = [{
+    Performs max-pooling operation on 3D input.
```
Done

Tessil commented on tensorflow/lite/kernels/internal/reference/pooling3d.h:
```diff
@@ -0,0 +1,193 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
PR was created in 2021, updated the copyright to 2022.

Tessil commented on tensorflow/lite/kernels/internal/types.h:
```diff
@@ -1008,6 +1008,22 @@ struct LeakyReluParams {
   int32_t output_shift_identity;
 };
 
+struct Pool3DParams {
```
Done

Tessil commented on tensorflow/lite/tools/versioning/runtime_version.cc:
```diff
@@ -378,6 +378,8 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
               {{BuiltinOperator_GELU, 2}, "2.9.0"},
               {{BuiltinOperator_DYNAMIC_UPDATE_SLICE, 1}, "2.9.0"},
               {{BuiltinOperator_UNSORTED_SEGMENT_PROD, 1}, "2.10.0"},
+              {{BuiltinOperator_AVERAGE_POOL_3D, 1}, "2.10.0"},
```
Done

Tessil commented on tensorflow/lite/kernels/pooling3d_test.cc:
```diff
@@ -15,47 +15,52 @@ limitations under the License.
-    if (pool_type == kAverage) {
-      SetCustomOp("AveragePool3D", custom_option,
-                  ops::custom::Register_AVG_POOL_3D);
+    if (pool_type == kAveragePool) {
```
Done

Tessil commented on tensorflow/lite/testing/op_tests/pool3d.py:
```diff
@@ -62,28 +73,26 @@ def build_graph(parameters):
 @register_make_test_function()
 def make_avg_pool3d_tests(options):
-  make_pool3d_tests(tf.nn.avg_pool3d)(options, expected_tf_failures=6)
+  make_pool3d_tests(tf.nn.avg_pool3d)(options, expected_tf_failures=18)
```
Yes, the initial tests have 6 expected failures and we run the same tests for both int8 and int16, hence 18 expected failures.

Tessil commented:
Thank you for the review.

Dusted off the PR a bit, fixed the merge conflicts and addressed the comments. Only some MLIR tests still need to be added which I will do when I have the occasion.

Tessil committed:
Add MLIR tests for AVERAGE/MAX_POOL_3D

Tessil commented on tensorflow/compiler/mlir/lite/ir/tfl_ops.cc:
```diff
@@ -3735,6 +3735,21 @@ int64_t AveragePool2DOp::GetArithmeticCount(Operation *op) {
   return -1;
 }
 
+//===----------------------------------------------------------------------===//
```
Done

Tessil commented on tensorflow/compiler/mlir/lite/ir/tfl_ops.td:
```diff
@@ -5517,6 +5517,71 @@ No value constant op.
   }];
 }
 
+def TFL_AveragePool3DOp: TFL_Op<"average_pool_3d", [
```
Done

Tessil commented on tensorflow/compiler/mlir/lite/transforms/legalize_patterns.td:
```diff
@@ -566,3 +570,34 @@ def LegalizeMultinomial : Pat<
   (TF_XlaDynamicUpdateSliceOp $operand, $update, $start_indices),
   (TFL_DynamicUpdateSliceOp $operand, $update, $start_indices)>;
+
+def LegalizeAveragePool3D : Pat<(TF_AvgPool3DOp $input,
```
Done

Tessil commented:
@sirakiin I added the MLIR tests, if you could check if everything looks good. Thanks

sirakiin reviewed (approved):
Looks good, many thanks for the contribution!

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
@gbaned @sirakiin Thanks, I fixed the conflict. The PR will need re-approval.

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
@gbaned Thanks, I fixed the conflicts @sirakiin It'll need re-approval. Thanks!

abattery commented:
@alankelly any opinions on this operator addition?

Tessil committed:
Replace the deprecated NoSideEffect trait by Pure

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil committed:
Adapt to prefixed accessors change in tfl_ops

Tessil commented:
@gbaned Thanks, I fixed the conflicts.

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
@Tessil Any update on this PR? Please. Thank you!

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
@gbaned Thanks, I fixed the conflict.

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
Thanks, solved them.

alankelly commented on tensorflow/lite/kernels/internal/reference/pooling3d.h:
```diff
@@ -0,0 +1,193 @@
+                for (int filter_d3 = filter_d3_start; filter_d3 < filter_d3_end;
+                     ++filter_d3) {
+                  const int in_d3 = in_d3_origin + filter_d3;
+                  total += input_data[Offset(input_shape, batch, in_d1, in_d2,
```
Can you replace Offset with the partial calculation of the input index in each of the nested loops? Offset has terrible performance and I am working on getting rid of it.

Tessil replied:
I moved the implementations from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/pooling3d.cc to their own `pooling3d.h` file. 

As the PR has been open since more than one year and it's an important op to have for us, can we do it in two steps to avoid complexifying this PR more than it's currently? Once merged, I can create an optimisation PR to remove the `Offset` usage if it's alright to you.

Thanks

alankelly replied:
OK for me. Thanks

alankelly commented on tensorflow/lite/kernels/internal/reference/pooling3d.h:
```diff
@@ -0,0 +1,193 @@
+            T average = RoundAndAverage<T, ActivationT>(total, filter_count);
+            average = std::max<T>(average, activation_min);
+            average = std::min<T>(average, activation_max);
+            output_data[Offset(output_shape, batch, out_d1, out_d2, out_d3,
```
Same

alankelly commented on tensorflow/lite/kernels/internal/reference/pooling3d.h:
```diff
@@ -0,0 +1,193 @@
+                     ++filter_d3) {
+                  const int in_d3 = in_d3_origin + filter_d3;
+                  max =
+                      std::max(max, input_data[Offset(input_shape, batch, in_d1,
```
Same

alankelly commented on tensorflow/lite/kernels/internal/reference/pooling3d.h:
```diff
@@ -0,0 +1,193 @@
+            }
+            max = std::max<T>(max, activation_min);
+            max = std::min<T>(max, activation_max);
+            output_data[Offset(output_shape, batch, out_d1, out_d2, out_d3,
```
Same

alankelly commented on tensorflow/lite/tools/versioning/runtime_version.cc:
```diff
@@ -391,7 +391,9 @@ std::string FindMinimumRuntimeVersionForOp(tflite::BuiltinOperator op_code,
-           {{BuiltinOperator_SIGN, 2}, "2.12.0"}});
+           {{BuiltinOperator_SIGN, 2}, "2.12.0"},
+           {{BuiltinOperator_AVERAGE_POOL_3D, 1}, "2.11.0"},
+           {{BuiltinOperator_MAX_POOL_3D, 1}, "2.11.0"}});
```
2.12.0

Tessil replied:
Changed it, thanks.

alankelly replied:
Thanks

alankelly reviewed (commented):
I'm OK with the kernel changes. Adding @arfaian for the converter changes

Tessil committed:
Update runtime version for AVG/MAX_POOL_3D

jpienaar commented on tensorflow/compiler/mlir/lite/utils/validators.cc:
```diff
@@ -91,6 +91,20 @@ bool TFIntListIs1XYZ1(Operation *op, StringRef name, IntegerAttr *x,
+  const auto &elements = attr.cast<ArrayAttr>().getValue();
+  if (elements.size() != 5 ||
+      std::any_of(elements.begin(), elements.end(),
+                  [](Attribute e) { return !e.isa<IntegerAttr>(); }))
```
What can be it be except an IntegerAttr? I ask as in line 96 there is assumption that this is only used where ArrayAttr is already verified to be the attribute type and the function name says it should only be used for int lists. So seems like one assumption is checked while the other is assumed (and cast will just assert in debug mode).

jpienaar commented on tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir:
```diff
@@ -127,6 +127,28 @@ func.func @avgPool2D(%arg0: tensor<1x6x6x16xf32>) -> tensor<1x1x1x16xf32> {
+  // Unsupported strides
+  %3 = "tf.AvgPool3D"(%arg0) {T = "tfdtype$DT_FLOAT", data_format = "NDHWC", ksize = [1, 2, 3, 6, 1], padding = "VALID", strides = [1, 2, 3, 1, 3]} : (tensor<1x6x6x6x16xf32>) -> tensor<1x1x1x1x16xf32>
+
+  %5 = arith.addf %0, %1 : tensor<1x1x1x1x16xf32>
```
I'm assuming you are doing this just to avoid dead code elimination? If so, how about using tf.AddN with a comment.

Tessil replied:
From my understanding it was done to make it easier to have a single return tensor without having to mess with shapes. It's done in a similar way for 2D average and max pooling tests.

jpienaar commented on tensorflow/compiler/mlir/lite/tests/ops.mlir:
```diff
@@ -678,6 +678,66 @@ func.func @testMaxPool2DWrongOperandStorageType(tensor<1x7x7x16x!quant.uniform<i
 
+// CHECK-LABEL: testMaxPool3D
+func.func @testMaxPool3D(tensor<256x32x32x32x3xf32>) -> tensor<?xf32> {
+^bb0(%arg0: tensor<256x32x32x32x3xf32>):
```
This is the "old" form, you can just roll the bb args into the function signature as you did in the other test.

Tessil replied:
Done

jpienaar reviewed (commented):
Moving to builtin is mostly question for TFLite team and seems they are happy with it. Gave the converter parts and scan and seems reasonable (adding patterns to fuse into the op can be done as follow up).

Tessil committed:
Pass the arguments in the function signature of the MLIR tests instead of using the old ^bb form

Tessil committed:
Use tf.AddN instead of multiple arith.addf

Tessil committed:
Move CHECKs at the start of the tested op

Tessil commented on tensorflow/compiler/mlir/lite/tests/legalize-tf.mlir:
```diff
@@ -127,6 +127,28 @@ func.func @avgPool2D(%arg0: tensor<1x6x6x16xf32>) -> tensor<1x1x1x16xf32> {
+  func.return %7 : tensor<1x1x1x1x16xf32>
+
+// CHECK-LABEL: func @avgPool3D
+// CHECK:  "tfl.average_pool_3d"(%arg0) {filter_depth = 2 : i32, filter_height = 3 : i32, filter_width = 6 : i32, fused_activation_function = "NONE", padding = "VALID", stride_d = 2 : i32, stride_h = 3 : i32, stride_w = 1 : i32} : (tensor<1x6x6x6x16xf32>) -> tensor<1x1x1x1x16xf32>
```
Done (note that it was done to be similar to other surrounding tests, I can change it back for coherence if needed).

Tessil commented:
@gbaned @jpienaar Thanks for the review, I addressed the comments and updated the PR.

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
@gbaned Done, thanks

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

gbaned commented:
Hi @sirakiin Can you please review this PR ? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

Tessil commented:
@gbaned @sirakiin Would it be possible to check what is blocking this PR internally as it has been open for 18 months and approved two times before? We can eventually close this PR and leave it internally to you?

gbaned commented:
Hi @Tessil Can you please resolve conflicts? It will help us to check the internal errors. Thank you!

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/builtin_average_max_pooling_3d

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @Tessil  Any update on this PR? Please. Thank you!

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

Tessil commented:
Thanks I fixed the conflicts.

Though with the new 'BuiltinOptions2`, we have some problems with the generated `tensorflow/compiler/mlir/lite/operator_converters.inc` I'm not sure how to solve. @alankelly could you eventually take a look when you have some time? Thanks!

gbaned commented:
Hi @Tessil Still, conflicts are appearing. Can you please resolve those? Thank you!

Tessil committed:
Merge branch 'master' into toupstream/builtin_average_max_pooling_3d

gbaned commented:
Hi @alankelly Can you please review this PR ? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @Tessil Can you please resolve conflicts? Thank you!

damiles commented:
> Hi @Tessil Can you please resolve conflicts? Thank you!

I did a MR to @Tessil (https://github.com/Tessil/tensorflow/pull/1) to solve the conflict, but looks like he is not receiving notifications :'(

gbaned commented:
Hi @Tessil I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! Thank you for your contribution.

Tessil commented:
@gbaned Sorry for the late reply, we can leave the PR as closed for now. It has been open for more than two years and we don't have enough time to maintain it currently.

damiles commented:
If it's possible, I want to maintain it. This feature is needed for some projects that I'm working on.

Tessil commented:
@damiles Sorry for the delay. Feel free to open a new PR using the code of this one as long as the commit history of this PR is preserved. It might be good to check the interest of the TensorFlow team in this feature beforehand though.
